### PR TITLE
Add terraform module

### PR DESCRIPTION
**Pull request by sed-i**
## Solution

TODO:
- [x] Call `terraform fmt -check -recursive` as part of CI
  - [x] github workflow (local demo)
  - [ ] Tandem PR in observability repo
- [x] Add `.tftest.hcl` files to use with `terraform test` as part of itests.
  - [x] [Use `local-exec`](https://github.com/canonical/terraform-modules/blob/b2626da89052792da7a96a304fed233ed1a55371/modules/machine/postgresql-ha/main.tf#L23) to create the s3 bucket.
  - [ ] Include OpenTofu in itests (?)
  - [ ] See if we want to use [pytest-terraform](https://pypi.org/project/pytest-terraform/) instead of `terraform test`.
- [x] Add integrations into module itself (is it possible / sensible?)
- [x] Decide where (which repo) to keep the worker module, the coordinator module, and potentially the mimir-ha module.
- [x] Add a "[constraints](https://registry.terraform.io/providers/juju/juju/latest/docs/resources/application#constraints)" input var. [Example](https://discourse.charmhub.io/t/pod-priority-and-affinity-in-juju-charms/4091/13).

Refs:
- https://discourse.charmhub.io/t/introducing-the-juju-terraform-provider/6939
- https://github.com/juju/terraform-provider-juju/blob/main/.github/workflows/test_integration.yml
- https://registry.terraform.io/providers/juju/juju/latest/docs#example-usage
- https://github.com/canonical/terraform-modules/tree/main/modules/machine/postgresql-ha

## Testing Instructions
```
terraform test
```

### CODE REVIEW

**Comment by simskij** [comment]
Why is this in the worker repo? 😅

**Review by marcoppenheimer — COMMENTED** [review]
I don't see any major issues, looks great!
Though, I'm hesitant about splitting the testing frameworks to include Terraform-specific tests for 'charm module' testing - the value-add over pytest-operator `wait_for_idle(status="active")` is unclear, as it's really just `juju deploy app`.

Although it's really old and probably a bit rubbish, we're doing the following in the [Kafka bundle](https://github.com/canonical/kafka-bundle/blob/main/tests/integration/terraform/test_terraform.py) to just use ops-test, so that it's part of the existing test suite we'd find in the charm itself.

In the 'solutions' repo, I think the Terraform tests have a lot more value! Though I'd like for these to be implemented by Solutions QA (or equivalent party) in some automated way, so that we're testing the integrations across multiple products/resources.
And having these 'wide integration' tests being managed not by us can help inform teams when new releases break compatibility with others (e.g TLS releases...)

File: terraform/main.tf [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  model = var.model_name
+  trust = var.trust
+
+  charm {
+    name    = "mimir-worker-k8s"
+    channel = var.channel
+  }
+  units  = var.units
+  config = var.config
+  # TODO: add revision, if given
```
Comment (orfeas-k) [inline terraform/main.tf:13]: Looking at juju provider [docs](https://registry.terraform.io/providers/juju/juju/latest/docs/resources/application#nested-schema-for-charm), revision is to be added under `charm{}`. Maybe including the `TODO` there would be more clear

File: terraform/variables.tf [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+variable "channel" {
+  description = "Charm channel"
+  type        = string
+  default     = "latest/stable"
+}
+
+variable "revision" {
+  description = "Charm revision"
+  type        = number
+  nullable    = true
```
Comment (orfeas-k) [inline terraform/variables.tf:20]: Can't this line be omitted? https://developer.hashicorp.com/terraform/language/values/variables#disallowing-null-input-values

File: terraform/main.tf (hunk 1 above)
Comment (sed-i) [reply terraform/main.tf]: Added :)

File: terraform/variables.tf (hunk 2 above)
Comment (sed-i) [reply terraform/variables.tf:20]: Yes, `true` is default, but it's so weird to me that I though keeping it explicit would be valuable.

File: terraform/variables.tf [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+variable "model_name" {
+  description = "Model name"
+  type        = string
+}
+
+variable "channel" {
+  description = "Charm channel"
+  type        = string
+  default     = "latest/stable"
+}
```
Comment (orfeas-k) [inline terraform/variables.tf:15]: The TF style guide suggests this file keeps variables in alphabetical order (ref https://developer.hashicorp.com/terraform/language/style#file-names)

File: terraform/outputs.tf [hunk 4]
```diff
@@ -0,0 +1,9 @@
+output "app_name" {
+  value = juju_application.mimir_worker.name
+}
+
+output "requires" {
+  value = {
+    mimir_cluster = "mimir-cluster"
+  }
+}
```
Comment (ghislainbourgeois) [inline terraform/outputs.tf:9]: I like this naming, I see that you added this to the spec.

File: terraform/variables.tf [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  description = "Number of units"
+  type        = number
+  default     = 1
+}
+
+variable "trust" {
+  description = "Equiv of juju deploy --trust"
+  type        = bool
+  default     = false
+}
```
Comment (ghislainbourgeois) [inline terraform/variables.tf:40]: As I said in the spec discussion, I am not sure having this as an input makes sense. In my experience, the charm either requires it or not. Having this as an input would only allow broken configuration in that case. Are you aware of charms that can run without `trust`, but that can also benefit from it?

File: terraform/tests/minimal/main.tf
Comment (ghislainbourgeois) [inline terraform/tests/minimal/main.tf:1]: I am not convinced that tests at the charm module level make a lot of sense, unless we replace the integration tests of the charm with those through Terraform.

To me, they look like mostly duplication, and are less readable than the standard integration tests.

**Review by ghislainbourgeois — COMMENTED** [review]
This seems like a good example for the spec. Thanks for doing this.

File: .github/workflows/pull-request.yaml [hunk 6]
```diff
@@ -10,3 +10,12 @@ jobs:
     name: PR
     uses: canonical/observability/.github/workflows/charm-pull-request.yaml@main
     secrets: inherit
+
```
Comment (michaeldmitry) [inline .github/workflows/pull-request.yaml:13]: CI is in https://github.com/canonical/observability/pull/209

File: terraform/variables.tf [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  description = "Config options as in the ones we pass in juju config"
+  type        = map(string)
+  default     = {}
+}
+
+# We use constraints to set AntiAffinity in K8s
+# https://discourse.charmhub.io/t/pod-priority-and-affinity-in-juju-charms/4091/13?u=jose
+variable "constraints" {
+  description = "Constraints to be applied"
+  type        = string
```
Comment (michaeldmitry) [inline terraform/variables.tf:22]: should we add a default value?

**Review by sed-i — COMMENTED** [review]
Tested indirectly by deploying https://github.com/canonical/observability/pull/212.

File: terraform/versions.tf [hunk 8]
```diff
@@ -0,0 +1,9 @@
+terraform {
+  required_version = ">= 1.5"
+  required_providers {
+    juju = {
+      source  = "juju/juju"
+      version = "~> 0.14.0"
```
Comment (MichaelThamm) [inline terraform/versions.tf:6]: Not sure what we should standardise our .gitignores to for TF but I found this from one of our spec examples:

https://github.com/canonical/grafana-k8s-operator/pull/350/files#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947

There are a lot more TF files created than just in `.terraform/`

**Comment by MichaelThamm** [comment]
Please add .gitignore changes for common TF files we do not want in git

**Review by MichaelThamm — APPROVED** [review]
Looks good to me!

**Pull request merged**